src/clients/BundleDataClient/BundleDataClient.ts

```diff
@@ -1106,72 +1114,186 @@ export class BundleDataClient {
+          if (deposit.originChainId !== originChainId || deposit.destinationChainId !== destinationChainId) {
+            return;
+          }
+          const isDuplicateDepositInBundle = bundleDepositHashes
+            .slice(0, currentBundleDepositHashIndex)
+            .some((_depositHash) => {
+              const { relayDataHash: _relayDataHash } = decodeBundleDepositHash(_depositHash);
+              return _relayDataHash === relayDataHash;
+            });
+          // Don't refund duplicate deposits from a prior bundle, as they should have been refunded already
+          // if they coincided with another deposit in the same bundle. If they didn't, then its input
+          // amount was used to refund a pre-fill.
+          // We will refund any duplicate deposits the first time that we see a deposit hash in this bundle.
+          // If this is the first time we are seeing this deposit hash, then refund any duplicate deposits since
+          // a fill exists for it and these duplicate deposits can no longer be refunded for expiry.
+          // This means unfortunately that every duplicate deposit that is sent that
+          // does not accompany another deposit in the same bundle will not be refunded. This should be unlikely.
+          // This rule also allows us to protect honest depositors who accidentally send duplicate deposits
+          // in rapid succession in most cases, unless they are unlucky enough to send duplicate deposits
+          // in different bundle block ranges.
+          const duplicateDepositsInBundle = deposits!.slice(index + 1);
+
+          // We are willing to refund a pre-fill multiple times for each duplicate deposit.
+          // This is because a duplicate deposit for a pre-fill cannot get
+          // refunded to the depositor anymore because its fill status on-chain has changed to Filled. Therefore
+          // any duplicate deposits result in a net loss of funds for the depositor and effectively pay out
+          // the pre-filler.
+
+          // If fill exists in memory, then the only case in which we need to create a refund is if the
+          // the fill occurred in a previous bundle. There are no expiry refunds for filled deposits.
+          if (fill) {
+            if (!isDuplicateDepositInBundle && fill.blockNumber < destinationChainBlockRange[0]) {
+              duplicateDepositsInBundle.forEach((duplicateDeposit) => {
+                updateExpiredDepositsV3(expiredDepositsToRefundV3, duplicateDeposit);
+              });
+              if (canRefundPrefills) {
+                // If fill is in the current bundle then we can assume there is already a refund for it, so only
+                // include this pre fill if the fill is in an older bundle. If fill is after this current bundle, then
+                // we won't consider it, following the previous treatment of fills after the bundle block range.
+                if (!isSlowFill(fill)) {
```

Comment:
Is this check necessary/Is it just to not bloat validatedBundleV3Fills? I notice that for the normal fill case, we do not check do this check before pushing to validatedBundleV3Fills since when we are modifying the refund dictionary, we exit early if the fill was a slow fill. (same question for this check on line 1244).

Reply:
Good catch! I agree with you

Reply:
https://github.com/across-protocol/sdk/pull/835/commits/dc6d4826d0fa465501113d022b73ba2727f5e9d8

Reply:
This was actually a crucial bug fix, because even slow fill executions add to the LP fees. This is how LP's get paid when slow leaves get executed and the LP fee is subtracted from the recipient's inputAmount. https://github.com/across-protocol/sdk/blob/13920ee180aa5946502cc27bbd2beeaf0046d1e3/src/clients/BundleDataClient/BundleDataClient.ts#L107

Reply:
Hm actually I'm thinking about this more we may need to keep these checks in. When a fill is a slow fill execution normally we'd want to charge lp fees on them. However if a slow fill leaf is being treated as a pre fill execution then we'd be charging lp fees a second time. This is because when the leaf execution originally happened, lp fees must have been charged. Because that fill would always be able to be matched with a deposit.